Funcky/Internal/Mixer.cs

```diff
@@ -1,19 +1,25 @@
+using System.Collections.Immutable;
+
 namespace Funcky.Internal
 {
     internal static class Mixer
     {
-        public static IEnumerable<T> ToRandomEnumerable<T>(IList<T> source, Random random)
+        public static IEnumerable<TSource> ToRandomEnumerable<TSource>(IList<TSource> source, Random random)
+            => Enumerable
+                .Range(0, source.Count)
+                .Aggregate(ImmutableList<TSource>.Empty, AggregateShuffle(source, random));
```

Comment:
Could this `Aggregate` be split into a `Select` and a `ToList` call?

Reply:
Actually you do not even need the `ToList` in such a case, because we only return an IEnumerable, so the `ToList `would look redundant, however without the ToList multiple enumeration will most likely fail, because the source is mutated.
To make this clear, I would use Materialize instead.

But since the source is mutated and we deal with side effects here, I am pretty sure the Aggregate is safer and less confusing. 

I did not push my exercise, it would look like this:

```cs
public static IEnumerable<TSource> ToRandomEnumerable<TSource>(IList<TSource> source, Random random)
    => ProjectShuffledEnumerable(source, random)
        .Materialize();

private static IEnumerable<TSource> ProjectShuffledEnumerable<TSource>(IList<TSource> source, Random random)
    => from currentIndex in Enumerable.Range(0, source.Count)
       let randomIndex = random.Next(currentIndex, source.Count)
       select UseElement(source, currentIndex, randomIndex);

private static TSource UseElement<TSource>(IList<TSource> source, int currentIndex, int randomIndex)
{
    var result = source[randomIndex];

    source[randomIndex] = source[currentIndex];

    return result;
}

```

Reply:
I'm convinced 🙂